lib/rmt/downloader.rb

```diff
@@ -56,25 +61,30 @@ def process_queue
       local_file = make_local_path(remote_file)
     end while File.exist?(local_file) # rubocop:disable Lint/Loop
 
-    # The request is wrapped into a fiber for exception handling
-    request_fiber = Fiber.new do
-      begin
-        make_request(remote_file, local_file, request_fiber, queue_item[:checksum_type], queue_item[:checksum])
-      rescue RMT::Downloader::Exception => e
-        @logger.warn("× #{File.basename(local_file)} - #{e}")
-      ensure
-        process_queue
+    was_deduplicated = deduplicate(queue_item[:checksum_type], queue_item[:checksum], local_file)
+
+    unless was_deduplicated
+      # The request is wrapped into a fiber for exception handling
+      request_fiber = Fiber.new do
+        begin
+          make_request(remote_file, local_file, request_fiber, queue_item[:checksum_type], queue_item[:checksum])
+        rescue RMT::Downloader::Exception => e
+          @logger.warn("× #{File.basename(local_file)} - #{e}")
+        ensure
+          process_queue
+        end
       end
+
+      @hydra.queue(request_fiber.resume)
     end
 
-    @hydra.queue(request_fiber.resume)
   end
 
   def verify_checksum(filename, checksum_type, checksum_value)
     hash_function = checksum_type.gsub(/\W/, '').upcase.to_sym
     hash_function = :SHA1 if (hash_function == :SHA)
 
-    unless (KNOWN_HASH_FUNCTIONS.include? hash_function)
+    unless KNOWN_HASH_FUNCTIONS.include? hash_function
```

Comment:
Why you hate parens :cry:

Reply:
I love them in other languages, but not in ruby. ^_^